### ISSUE TITLE
Range of hasNumericalUnit

### ISSUE THREAD

**Issue by rpgoldman** [HajoRijgersberg/OM#57]
I am looking at your README with the UML diagram for OM-2, and at the same time, I have the OM2 ontology open in Protege.

When I look at the UML diagram, I see that the range of `hasNumericalUnit` is illustrated as being of type `xml:float`.  However, when I look at the ontology in Protege, I don't see any range constraint on `hasNumericalUnit`, beyond it being a datatype property.

Below this, in the README, I see this:

```
ex:_10Centimetres rdf:type om:Measure ;
  om:hasNumericalValue "10"^^xsd:double ;
  om:hasUnit om:centimeter .
```

There are a number of problems with this that come from the fact that OWL's numeric datatypes (in my opinion) are bonkers (this is a technical term from logic 😉):

1. If we were to add a range constraint  `DataPropertyRange( om:hasNumericalValue xsd:float )`, then the OM ontology plus `ex:_10Centimetres` would be *inconsistent*, because `xsd:float` and `xsd:double` are *disjoint* types.  See [Datatype Maps in the OWL2 specs](https://www.w3.org/TR/owl2-syntax/#Datatype_Maps).
2. Even worse, there is *no* numeric type that brings together the types `xsd:float`, `xsd:double`, and `owl:real`.  This means that it is very easy to render an ontology inconsistent by, for example, supplying an integer value for a property that is typed as `xsd:float` or `xsd:double`.  Or, for that matter, supplying one of these two float types as the value of a property that has any type that is `owl:real` or any subtype thereof.

Someone I discussed this with says that it's a common standard practice to represent whole numbers using `xsd:integer` or a subtype thereof, and any quantity that might not be a whole number as an `xsd:decimal`.  This avoids the kind of issues I describe above.

There doesn't seem to be anything consistent with the OWL spec that would permit normal uses of `xsd:float` or `xsd:double` -- any use of these types seems likely to cause inconsistencies.

I conjecture that this is a bad consequence of adopting the XSD data types -- which are really *syntactic* into a logic that is meant to be semantic.

**Comment by rpgoldman** [comment]
Your statements clarify the situation, yes, and we can probably close this issue.

But it's clear that users need guidance when they supply values or apply constraints on measures.  If they assume that normal numerical conversions will apply, then very bad things will happen.

I don't have very good solutions to offer:

1. Make the suggestion that I refer to above: encourage users to always supply either an `xsd:decimal` or an `xsd:integer` everywhere.  Constrain `hasNumericalValue` to have a range of `xsd:decimal`.  This seems as good a fix as possible, but potentially breaks uses of OM.
2. We should get OWL reasoners that do something more sensible with numbers!  Even better, the OWL spec should be repaired where numerical data types are concerned.

**Comment by jakebeal** [comment]
Looking more closely into `xsd:decimal`, I'm concerned that constraining to `xsd:decimal` would create additional problems.
I notice two key issues in the definition of `xsd:decimal`:

1. Infinity, -Infinity, and NaN are not included
2. Precision cannot be specified, i.e., 2 = 2.0 = 2.00 _arguably a feature in the context of OWL, but not measurement_

Lack of precision isn't any worse than using an `xsd:float` or `xsd:double`. Lack of the infinities and NaN is a major issue, however, since _some instruments return those values_.  Thus, if I am using OM to encode information about the results of a physical measurement operation, `xsd:decimal` will prevent me from recording what the instrument returned and force the use of a different representation.

There does appear to be an XSD datatype that was specifically designed to address these issues, [xsd:precisionDecimal](https://www.w3.org/TR/xsd-precisionDecimal/), but it [did not make it into XML Schema 1.1](https://stackoverflow.com/questions/61349238/why-isnt-xsprecisiondecimal-part-of-the-xsd1-1) and as a consequence is not supported by libraries.

**Comment by rpgoldman** [comment]
@HajoRijgersberg Sorry to be ignorant, but could you point me in the right direction to investigate the distinction between *ideal measures* and *concrete measurements*?  There is likely a standard description of this distinction, but I don’t know the vocabulary and am unable to easily search for relevant literature.  Hence my misuse of the term “quantities” above.

My guess is that this terminology will be easy for you to specify. If not, please just close this.

Thanks!

**Comment by jakebeal** [comment]
@rpgoldman When thinking about measurements in biology, I am not typically concerned about the limited precision of an `xsd:float`, since that gives us seven decimal places of precision, and that's far more than any typical biolab instruments are able to give us. My baseline assumption for approach would then be one of the following two:
1. Convert everything into floats, or 
2. Pull at least the numerical aspects of reasoning out of OWL as it's not really suited for them anyway.

in either case leaving any requirement for additional precision as an exception case to be handled in the future.

**Comment by dr-shorthair** [comment]
RDFS and OWL are designed for logic, to support reasoning, not arithmetic computation. Numerics are not central, hence that was outsourced to XSD (and inherits flaws from there). 

But I repeat my point: it looks to me that you are overloading a datatype with information that should be explicitly modeled in the application.

**Comment by rpgoldman** [comment]
I can’t speak for Jake, but if one wants to be able to talk about the meaning of experimental measurements, one needs to be able to represent _both_ the meaning _and_ the measurements themselves, and to do that, one must be able to take into account the limitations of the measurements.

The decision to treat numerics as a side matter seems like a poor choice for a language to be used to talk about the scientific process as it is actually practiced.

**Comment by rpgoldman** [comment]
> > Another issue is that I don’t believe we can even mention something like a transcendental number, except by using an approximate value. On the one hand, that reflects what we can do in first order logic, but on the other hand it needlessly prevents us from using such values in an incomplete, but useful, way. E.g., talking about an angular value that is greater than pi.
> 
> Do you mean complex numbers? We got a topic about that too: #51.

No, actually I meant the real numbers that are *not* rational.  `xsd:decimal` is only sufficient to represent the rationals.  There's `owl:real`, but 

> The owl:real datatype does not directly provide any lexical forms. 

I'm not sure how one refers to a non-rational real since there are no lexical forms for it, and they aren't named individuals so I can't say `:pi a owl:real`.

I would think that there would be some discussion of this, but I haven't been able to find it.

**Comment by HajoRijgersberg** [comment]
I have thought about it. This all seems to fit under the subject 'lists of numbers' (or 'sets of numbers' - in Dutch we say 'getalverzamelingen'; it's unclear to me how it exactly translates to English). The complex numbers are special, since they are indeed represented by a tuple. But on the other hand - I agree with you - real numbers should in fact too, as a fraction, i.e. with a numerator and a denominator - also a kind of a tuple (in a way). Indeed, approximating real numbers does not seem sensible. So perhaps we should broaden the already existing issue 'Complex numbers' on this GitHub to 'Lists of numbers'. Does that seem sensible?

**Comment by rpgoldman** [comment]
I think it's a general problem of "semantics of numbers." For some reason, the OWL designers chose to represent numbers in a way that is *fundamentally* syntactic instead of semantic. We can only talk about digital encodings of numbers, not numbers in the abstract, as far as I can tell. Numbers encoded as floats are semantically different from the rationals, and numbers encoded as single precision floats are different from double precision floats. That's accurate, but it doesn't let us talk sensibly about denotations.

I suppose that permits some forms of sound and relatively efficient reasoning, but only at the cost of limiting us to the "Only Somewhat Semantic Web." Something incomplete but more expressive would be nice to have.

**Comment by HajoRijgersberg** [comment]
I fully agree with you. And it confronts us with a great problem, from an optimistic perspective a challenge, but I think it's huge.
Anyway, hope we can contribute through this GitHub in this quite fundamental issue.
Regarding tuples, Maxim van den Wynckel has just opened a new issue on this GitHub: https://github.com/HajoRijgersberg/OM/issues/63. Maybe this is the issue I suggested above.
To be continued, all!
